### DISCUSSION TITLE
How can I read back the content of a wgpu::Surface after a render?

### DISCUSSION THREAD

**Discussion by gracicot** [gpuweb/gpuweb#3420] [Q&A]
I'm a beginner using webgpu. I currently use the `webgpu_cpp.h` file provided by emscripten and by google dawn. I have a `wgpu::Surface` in my C++ code (`WGPUSurface` in C) that I created and my goal is to render a simple red triangle in a buffer then use the buffer data to do stuff on the CPU. For example, I would render such triangle in the buffer, and read back those values:

```
0 0 0 0 1 0 0 0
0 0 1 1 1 1 0 0
1 1 1 1 1 1 0 0
0 0 1 1 1 1 1 0
0 0 0 0 1 1 1 0
0 0 0 0 0 0 1 1
0 0 0 0 0 0 0 0
```

And read it back as an array (assuming a `r8unorm` format)

```c++
unsigned char* buffer = wgpuGetSurfaceContent(surface);

for (int i = 0; i < 8*8; ++i) {
    std::print("{} ", (int)buffer[i]); // expects 0 if black and 255 if red
}
```

Does such function exist and how do I synchronize the reading/writing? Thank you.

**Comment by Kangz** [comment]
Note that the WebGPU C API is not discussed here, a better place would be [webgpu-native,](https://github.com/webgpu-native/webgpu-headers) the [Dawn's communication channels](https://dawn.googlesource.com/dawn) or [wgpu](https://github.com/gfx-rs/wgpu).

That said, at the moment it is not possible to read back from the swapchain texture in the webgpu-native headers (though it might change). For now you should render to a texture outside of any swapchain and copy from it to a buffer, before mapping the buffer for reading. If you also need the texture on the screen, then you can copy from that outside texture to the swapchain texture with a manual blit.

In the future when the swapchain texture can potentially be `CopySrc` then you can do the same thing but without the "outside texture" mentioned above.

**Reply by kainino0x** [reply]
+1, **if** you don't need to display the rendering result, the appropriate way to get rendering results back to the CPU is to not use a Surface (equivalently, in JS, a canvas) at all. Instead, render to a plain texture made from `createTexture` with `RENDER_ATTACHMENT | COPY_SRC`, copy to a `COPY_DST | MAP_READ` buffer with `copyBufferToTexture`, and then map the buffer with `mapAsync(GPUMapMode.READ)`.

**Reply by gracicot** [reply]
I see. Thank you! I think I was mislead since in OpenGL you must create a framebuffer to render into, even if it is never displayed.

**Reply by kainino0x** [reply]
Makes sense! The equivalent of a GL framebuffer in WebGPU is essentially the `GPURenderPassDescriptor` passed to `beginRenderPass`, which doesn't need to be created ahead of time. There's additional stuff in there too, not relating to the framebuffer.
